readme.md

```diff
@@ -55,19 +55,22 @@ Options object with following parameters.
 
 ##### incrementer
 
-Type: `(filePath: string) => string`\
+Type: `(filePath: string) => [string, string]`\
 Default: parentheses incrementer (`file.txt` → `file (1).txt`)
 
 A simple function that accepts a file path, and increments its index. It's incrementer's responsibility to extract an already existing index from passed file.
 
+The incrementer has to return a tuple of `[originalFilename, incrementedFilename]`, where `originalFilename` is the filename without incrementation sequence, and `incrementedFilename` is a filename with input incrementation sequence bumped by one.
```

Comment:
Why does it have to return the original filename? That's not very user-friendly. It think we can solve this without forcing the user to return it.

Reply:
Yeah I don't like it much either, but if path passed to unusedFilename is already incremented, I thought it would be valuable to have the un-incremented version to add IDs to in case of a MaxTryError, no?

```ts
unusedFilename('rainbow (1).txt', {maxTries: 0}); // error.originalFilename === `rainbow.txt`
```

Reply:
> I thought it would be valuable to have the un-incremented version to add IDs to in case of a MaxTryError, no?

Yes, but these things are not mutually exclusive. You can still store the original path internally instead of forcing the user to deal with it. For example, in a `WeakMap`.

Reply:
But we can't retrieve the original path, because we don't know the incrementation format beforehand. Only incrementer knows about that, so it's the only source of this.

Reply:
Ok, I missed that you're passing in a filename with a number to `unusedFilename`. Is that really a behavior we want to support? And is it expected behavior?

I think users could easily assume that `unusedFilename('rainbow (1).txt');` would return `rainbow (1) (1).txt` if `rainbow (1).txt` exists. Because the assumption is that the filename passed to `unusedFilename` is the original.

Reply:
That's one of the main reasons for this PR. To prevent silly filenames like `rainbow (1) (1) (1) (1).txt`. I would honestly not expect a file name incrementer to behave like this.

> the assumption is that the filename passed to `unusedFilename` is the original

I made this PR because I'm making a user facing app where this assumption is often wrong.

For example, lets imagine we are imlementing a file saving dialogue, and browsing a website where every image has a filename `image.png` (ehm... Discord). Do you want the files on HDD be

```
image (33).png
```

or 

```
image (1) (1) (1) (1) (1) (1) (1) (1) (1) (1) (1) (1) (1) (1) (1) (1) (1) (1) (1) (1) (1) (1) (1) (1) (1) (1) (1) (1) (1) (1) (1) (1) (1).png
```

Or my particular use case: I have a several file based operations. Each one has a UI drop zone into which users drag and drop files to be processed. All operations try to not override the input by incrementing the filename of the result.

They drop the file into drop zone 1, they get `filename (1)`. They take the result, and drop it into another drop zone to run a different operation on it. Everyone would expect the app to continue incrementing, and not just append.

If we are going to just append, than what's the point of pretending we are going to incrementat in the first place? Lets just append UUID at the end of every passed file and be done with it.

Reply:
> That's one of the main reasons for this PR. To prevent silly filenames like rainbow (1) (1) (1) (1).txt. I would honestly not expect a file name incrementer to behave like this.

I guess we have different use-cases in mind. I don't think it's a common use-case to have the user re-run already incremented files. The common use-case is just wanting to save a file and ensuring it's not overwriting anything.

But sure, we can go with your version and support this. I think you should explicitly document that this is how it works though.